Libraries/Connect/Public/Interfaces/Streaming/AsyncAwait/ClientOnlyAsyncStreamInterface.swift

```diff
@@ -39,8 +39,9 @@ public protocol ClientOnlyAsyncStreamInterface<Input, Output> {
     /// - returns: An `AsyncStream` that contains all outputs/results from the stream.
     func results() -> AsyncStream<StreamResult<Output>>
 
-    /// Close the stream. No calls to `send()` are valid after calling `close()`.
-    func close()
+    /// Close the stream and await a response.
+    /// No calls to `send()` are valid after calling `closeAndReceive()`.
+    func closeAndReceive()
```

Comment:
In other Connect and gRPC operations, I'm used to seeing this atomically fuse the `close` and `receive` operations, so that this method returns the one and only response message (or an error). So this seems a little confusingly named since the caller still must call `receive` right after. Or am I misunderstanding this?

Reply:
Related to https://github.com/connectrpc/connect-swift/pull/271#discussion_r1619628243, there is no `receive` function on this interface (that's a function on the internal concrete implementation for wiring), and consumers do not manually fetch results from the stream. Consumers observe responses/results through the `results()` async stream, such as by doing `for await result in stream.results() { ... }`

Reply:
That said, I'm also open to better naming suggestions

Reply:
Alternatively, I am open to different naming suggestions. Also not opposed to changing this function to be `async`, removing `results()` from `ClientOnlyAsyncStreamInterface`, and returning the result here. @eseay thoughts?

Reply:
> Consumers observe responses/results through the `results()` async stream

If the server sends two messages for a client-stream operation, instead of only one, does the async stream send the first message to the application and _then_ send an error on receipt of the second message? If so, I think the appropriate behavior is to only deliver the message after trailers are received, so you know it was an OK status and exactly one message. That way the application only sees a response message for a well-formed response stream.

Maybe that's what already happens. Sorry if I'm making comments/nitpicks without actually having a clear understanding of what this is doing.

Reply:
> If the server sends two messages for a client-stream operation, instead of only one, does the async stream send the first message to the application and then send an error on receipt of the second message?

Yes, that is what happens now.

> If so, I think the appropriate behavior is to only deliver the message after trailers are received, so you know it was an OK status and exactly one message. That way the application only sees a response message for a well-formed response stream.

To verify I'm understanding correctly, you're suggesting the client pass headers but buffer the body until trailers are received, then verify that only one message was received and return only an error (and no messages) if multiple messages are present? If so, is this something the conformance tests can validate?

Reply:
> If so, is this something the conformance tests can validate?

Supposedly it is. It should be complaining if it sees a response message in the `ClientCompatResponse` but the test wasn't expecting one. Is it possible your conformance client is ignoring the initial received message when it subsequently encounters an error?

Reply:
Interesting. I added some logging and this is what I see when running the `Connect Unexpected Responses/HTTPVersion:1/TLS:false/client-stream/multiple-responses` test case:

```
Running 1 tests with reference server for server config {HTTP_VERSION_1, PROTOCOL_CONNECT, TLS:false}...

Received: headers(["content-type": ["application/connect+proto"], "content-length": ["65"]])


Converted into: headers(["content-type": ["application/connect+proto"], "content-length": ["65"]])


URLSession received 65 data bytes


Connect interceptor got 5 data bytes


Connect interceptor not passing message because it is empty


Connect interceptor got 5 data bytes


Connect interceptor not passing message because it is empty


Connect interceptor got 55 data bytes


Connect interceptor received end stream with 50 end stream response bytes


Received: complete(code: Connect.Code.ok, error: nil, trailers: nil)


Converted into: complete(code: Connect.Code.internalError, error: Optional(Connect.ConnectError(code: Connect.Code.unimplemented, message: Optional("unary stream has no messages"), exception: nil, details: [], metadata: [:])), trailers: nil)

Total cases: 1
1 passed, 0 failed
```

So I think what's happening is the Connect/gRPC/gRPC-Web interceptors are are stripping empty messages rather than passing them through to the end client ([here, for example](https://github.com/connectrpc/connect-swift/pull/271/files#diff-2cd1fd07672849e8274188b07c522f3994e56955e376b49af94c73ade98c74dbR217)) and the conformance suite is sending empty messages, so you're right that no message is being returned to the conformance runner. In fact, this is actually falling into the `unary stream has no messages` case because none of the empty messages are being passed through, and the test is passing because both happen to require the same status code.

What should the behavior be in this case? Should the client generally be passing empty messages through?

Reply:
Yes, an empty message is a valid message as long as the codec accepts an empty bytes input. If the codec does not accept the empty bytes, then you'd report the error to the client like any other error thrown by the decoding (like from a malformed paylad).

Reply:
Thanks for confirming. I pushed 2 commits:

The first one gets these two tests to fail by no longer stripping empty responses:
- `FAILED: Connect Unexpected Responses/HTTPVersion:1/TLS:false/client-stream/multiple-responses`
- `FAILED: gRPC-Web Unexpected Responses/HTTPVersion:1/TLS:false/trailers-in-body/client-stream/multiple-responses`

The second commit applies buffering to wait for client streams to complete before doing validation and then passing the results to the caller.